lib/include/pl/core/ast/ast_node_match_statement.hpp

```diff
@@ -0,0 +1,156 @@
+#pragma once
+
+#include <pl/core/ast/ast_node.hpp>
+#include <pl/core/ast/ast_node_literal.hpp>
+#include <pl/core/ast/ast_node_mathematical_expression.hpp>
+#include <utility>
+
+namespace pl::core::ast {
+
+    struct MatchCase {
+        std::unique_ptr<ASTNode> condition;
+        std::vector<std::unique_ptr<ASTNode>> body;
+
+        MatchCase() = default;
+
+        MatchCase(std::unique_ptr<ASTNode> condition, std::vector<std::unique_ptr<ASTNode>> body)
+            : condition(std::move(condition)), body(std::move(body)) { }
+
+        MatchCase(const MatchCase &other) {
+            this->condition = other.condition->clone();
+            for (auto &statement : other.body)
+                this->body.push_back(statement->clone());
+        }
+
+        MatchCase(MatchCase &&other) noexcept {
+            this->condition = std::move(other.condition);
+            this->body = std::move(other.body);
+        }
+
+        MatchCase &operator=(const MatchCase &other) {
+            this->condition = other.condition->clone();
+            for (auto &statement : other.body)
+                this->body.push_back(statement->clone());
+            return *this;
+        }
+    };
+
+    class ASTNodeMatchStatement : public ASTNode {
+
+    public:
+        explicit ASTNodeMatchStatement(std::vector<MatchCase> cases, std::optional<MatchCase> defaultCase)
+            : ASTNode(), m_cases(std::move(cases)), m_defaultCase(std::move(defaultCase)) { }
+
+        ASTNodeMatchStatement(const ASTNodeMatchStatement &other) : ASTNode(other) {
+            for (auto &matchCase : other.m_cases)
+                this->m_cases.push_back(matchCase);
+            other.m_defaultCase ? this->m_defaultCase.emplace(*other.m_defaultCase) : this->m_defaultCase = std::nullopt;
+        }
+
+        [[nodiscard]] std::unique_ptr<ASTNode> clone() const override {
+            return std::unique_ptr<ASTNode>(new ASTNodeMatchStatement(*this));
+        }
+
+        [[nodiscard]] std::vector<std::shared_ptr<ptrn::Pattern>> createPatterns(Evaluator *evaluator) const override {
+            evaluator->updateRuntime(this);
+
+            auto &scope = *evaluator->getScope(0).scope;
+            auto body  = getCaseBody(evaluator);
+
+            if(!body) return {};
+
+            for (auto &node : *body) {
+                auto newPatterns = node->createPatterns(evaluator);
+                for (auto &pattern : newPatterns) {
+                    scope.push_back(std::move(pattern));
+                }
+
+                if (evaluator->getCurrentControlFlowStatement() != ControlFlowStatement::None)
+                    break;
+            }
+
+            return {};
+        }
+
+         FunctionResult execute(Evaluator *evaluator) const override {
+            evaluator->updateRuntime(this);
+
+            auto body = getCaseBody(evaluator);
+
+            if(!body) return std::nullopt;
+
+            auto variables     = *evaluator->getScope(0).scope;
```

Comment:
You're calling `evaluator->getScope(0)` quite often here. Consider getting a reference to it at the start and then just use that